test/indexing.jl

```diff
@@ -0,0 +1,130 @@
+    @test @inferred(ArrayInterface.to_indices(a, ([true, true], :))) == (Base.LogicalIndex(Bool[1, 1]), Base.Slice(1:2))
+    @test @inferred(ArrayInterface.to_indices(a, (CartesianIndices((1,)), 1))) == (1:1, 1)
+    @test @inferred(ArrayInterface.to_indices(a, (1, 1, 1))) == (1,1, 1)
+    @test @inferred ArrayInterface.to_indices(a, ([CartesianIndex(1,1,1), CartesianIndex(1,2,1)],)) == (CartesianIndex{3}[CartesianIndex(1, 1, 1), CartesianIndex(1, 2, 1)],)
+    @test @inferred ArrayInterface.to_indices(a, ([CartesianIndex(1,1), CartesianIndex(1,2)],1:1)) == (CartesianIndex{2}[CartesianIndex(1, 1), CartesianIndex(1, 2)], 1:1)
+    @test_throws ErrorException ArrayInterface.to_indices(ones(2,2,2), (1, 1))
+
+end
+
+@testset "1-dimensional" begin
+    for i = 1:3
+        @test ArrayInterface.getindex(LinearIndices((3,)), i) == i
+        @test ArrayInterface.getindex(CartesianIndices((3,)), i) == CartesianIndex(i,)
+    end
+    @test ArrayInterface.getindex(LinearIndices((3,)), 2,1) == 2
+    @test ArrayInterface.getindex(LinearIndices((3,)), [1]) == [1]
+    # !!NOTE!! this is different than Base.getindex(::LinearIndices, ::AbstractUnitRange)
+    # which returns a UnitRange. Instead we try to preserve axes if at all possible so the
+    # values are the same but it's still wrapped in LinearIndices struct
+    @test ArrayInterface.getindex(LinearIndices((3,)), 1:2) == 1:2
+    @test ArrayInterface.getindex(LinearIndices((3,)), 1:2:3) === 1:2:3
+    @test_throws BoundsError ArrayInterface.getindex(LinearIndices((3,)), 2:4)
+    @test_throws BoundsError ArrayInterface.getindex(CartesianIndices((3,)), 2, 2)
+    #   ambiguity btw cartesian indexing and linear indexing in 1d when
+    #   indices may be nontraditional
+    @test_throws ArgumentError Base._sub2ind((1:3,), 2)
+    @test_throws ArgumentError Base._ind2sub((1:3,), 2)
+end
+
+@testset "2-dimensional" begin
+    k = 0
+    cartesian = CartesianIndices((4,3))
+    linear = LinearIndices(cartesian)
+    for j = 1:3, i = 1:4
+        k += 1
+        @test ArrayInterface.getindex(linear, i, j) == ArrayInterface.getindex(linear, k) == k
+        @test ArrayInterface.getindex(cartesian, k) == CartesianIndex(i,j)
+        @test ArrayInterface.getindex(LinearIndices(map(Base.Slice, (0:3,3:5))), i-1, j+2) == k
+        @test ArrayInterface.getindex(CartesianIndices(map(Base.Slice, (0:3,3:5))), k) == CartesianIndex(i-1,j+2)
+    end
+    @test linear[linear] == linear
+    @test linear[vec(linear)] == vec(linear)
+    @test linear[cartesian] == linear
+    @test linear[vec(cartesian)] == vec(linear)
+    @test cartesian[linear] == cartesian
+    @test cartesian[vec(linear)] == vec(cartesian)
+    @test cartesian[cartesian] == cartesian
+    @test cartesian[vec(cartesian)] == vec(cartesian)
+    @test linear[2:3] === 2:3
+    @test linear[3:-1:1] === 3:-1:1
+    @test_throws BoundsError linear[4:13]
+end
+
+@testset "3-dimensional" begin
```

Comment:
Mind adding some `@inferred` for good measure? Sometimes inference can hit heuristic limits.